subprojects/kotlin-dsl-provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/DefaultProjectSchemaProvider.kt

```diff
@@ -106,19 +112,50 @@ fun targetSchemaFor(target: Any, targetType: TypeOf<*>): TargetTypedSchema {
 }
 
 
-private
-fun accessibleExtensionsSchema(extensionsSchema: ExtensionsSchema) =
-    extensionsSchema.filter { isPublic(it.name) }
-
-
 private
 fun accessibleConventionsSchema(plugins: Map<String, Any>) =
     plugins.filterKeys(::isPublic).mapValues { inferPublicTypeOfConvention(it.value) }
 
 
 private
 fun accessibleContainerSchema(collectionSchema: NamedDomainObjectCollectionSchema) =
-    collectionSchema.elements.filter { isPublic(it.name) }
+    collectionSchema.elements
+        .filter { isPublic(it.name) }
+        .map(NamedDomainObjectSchema::toFirstKotlinPublicOrSelf)
+
+
+private
+fun NamedDomainObjectSchema.toFirstKotlinPublicOrSelf() =
+    publicType.concreteClass.kotlin.let { kotlinPublicType ->
+        takeIf { kotlinPublicType.visibility == KVisibility.PUBLIC }
```

Comment:
:x: The flow was throwing me off a little so I tried to rewrite it, in doing so I thought the distinction between Java public and Kotlin public deserved a mention so how about something like:

```kotlin


private
fun NamedDomainObjectSchema.toFirstKotlinPublicOrSelf(): NamedDomainObjectSchema =
    publicType.concreteClass.kotlin.let { kotlinType ->
        /**
         * Because a public Java class might not correspond necessarily to a
         * public Kotlin type due to Kotlin `internal` semantics, we check
         * whether the public Java class is also the first public Kotlin type,
         * otherwise we compute a new schema entry with the correct
         * Kotlin type.
         **/
        val firstPublicKotlinType = kotlinType.firstKotlinPublicOrSelf
        when {
            firstPublicKotlinType === kotlinType -> this
            else -> ProjectSchemaNamedDomainObjectSchema(
                name,
                firstPublicKotlinType.asTypeOf()
            )
        }
    }


private
fun KClass<out Any>.asTypeOf() = TypeOf.typeOf(java)
```

Reply:
Very good point, done. Thanks!